distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerProcess.java

```diff
@@ -0,0 +1,266 @@
+     *
+     * <p> Note that if {@link #detach()} has been called, this method is a no-op.
+     */
+    public synchronized void stop() {
+        if (detached) {
+            return;
+        }
+
+        sendShutdownMarker();
+        waitFor();
+    }
+
+    private static void sendArgs(ServerArgs args, OutputStream processStdin) {
+        // DO NOT close the underlying process stdin, since we need to be able to write to it to signal exit
+        var out = new OutputStreamStreamOutput(processStdin);
+        try {
+            args.writeTo(out);
+            out.flush();
+        } catch (IOException ignore) {
+            // A failure to write here means the process has problems, and it will die anyways. We let this fall through
+            // so the pump thread can complete, writing out the actual error. All we get here is the failure to write to
+            // the process pipe, which isn't helpful to print.
+        }
+        args.keystorePassword().close();
+    }
+
+    private void sendShutdownMarker() {
+        try {
+            OutputStream os = jvmProcess.getOutputStream();
+            os.write(BootstrapInfo.SERVER_SHUTDOWN_MARKER);
+            os.flush();
+        } catch (IOException e) {
+            // process is already effectively dead, fall through to wait for it, or should we SIGKILL?
+        }
+    }
+
+    private static Process createProcess(
+        ProcessInfo processInfo,
+        Path configDir,
+        Path pluginsDir,
+        OptionsBuilder optionsBuilder,
+        ProcessStarter processStarter
+    ) throws InterruptedException, IOException, UserException {
+        Map<String, String> envVars = new HashMap<>(processInfo.envVars());
+        Path tempDir = setupTempDir(processInfo, envVars.remove("ES_TMPDIR"));
+        if (envVars.containsKey("LIBFFI_TMPDIR") == false) {
+            envVars.put("LIBFFI_TMPDIR", tempDir.toString());
+        }
+
+        List<String> jvmOptions = optionsBuilder.getJvmOptions(configDir, pluginsDir, tempDir, envVars.remove("ES_JAVA_OPTS"));
+        // also pass through distribution type
+        jvmOptions.add("-Des.distribution.type=" + processInfo.sysprops().get("es.distribution.type"));
+
+        Path esHome = processInfo.workingDir();
+        Path javaHome = PathUtils.get(processInfo.sysprops().get("java.home"));
+        List<String> command = new ArrayList<>();
+        boolean isWindows = processInfo.sysprops().get("os.name").startsWith("Windows");
+        command.add(javaHome.resolve("bin").resolve("java" + (isWindows ? ".exe" : "")).toString());
+        command.addAll(jvmOptions);
+        command.add("-cp");
+        // The '*' isn't allowed by the windows filesystem, so we need to force it into the classpath after converting to a string.
+        // Thankfully this will all go away when switching to modules, which take the directory instead of a glob.
+        command.add(esHome.resolve("lib") + (isWindows ? "\\" : "/") + "*");
```

Comment:
Trivially, this could simplified to use java.io.File.separator

Reply:
I don't think we can! We specifically use the local isWindows here so that we can check the unix and linux behavior in tests, regardless of which platform we are actually on. See `ServerProcessTests.testCommandLine`